subsys/net/l2/openthread/openthread.c

```diff
@@ -246,13 +259,55 @@ static void openthread_process(void *context, void *arg2, void *arg3)
 {
 	struct openthread_context *ot_context = context;
 
+		// not restarting the joiner before a changing delay has passed
+		#define JOINER_DELAY_INITIAL 2000
+		#define JOINER_DELAY_MAX 32000
+		// if no join in 10 minutes, reboot with better luck
+		#define JOINER_REBOOT_TIME K_MINUTES(10)
+		s64_t joiner_delay_ms = JOINER_DELAY_INITIAL;
+		s64_t reboot_ref_ms = k_uptime_get();
+
+
 	while (1) {
 		while (otTaskletsArePending(ot_context->instance)) {
 			otTaskletsProcess(ot_context->instance);
 		}
 
 		otSysProcessDrivers(ot_context->instance);
 
+		// restart joiner if
+		// - in autostart mode and join hasn't succeeded yet
+		// - joiner is back to idle
+		if (joiner_autostart_started && !joiner_successful
+			&& otJoinerGetState(ot_context->instance)
+			== OT_JOINER_STATE_IDLE) {
+			if (k_uptime_get() - reboot_ref_ms >
+				JOINER_REBOOT_TIME) {
+				// printk doesn't seem to do the job
+				printf("Joining failed for %d seconds..."
+					   "REBOOTIING now!\n\n",
+					   JOINER_REBOOT_TIME / 1000);
+				sys_reboot(SYS_REBOOT_COLD);
+			}
+			//delay before restarting joiner
+			k_sleep(joiner_delay_ms);
```

Comment:
If the value is in ms, then use `k_msleep()` instead

Reply:
We definitely should not block the OpenThread processing thread. In general, I don't think we should keep this logic in the main loop.

OpenThread calls `ot_joiner_start_handler` function when joining procedure is complete (successful or not). In case we need to retry the procedure, we should keep the retry logic in there. If the retry attempt has to be delayed, we should schedule a delayed work.

And in general, the retry-parameters should be configurable (i.e. initial delay, consider replacing max delay and reboot time with maximum retry count).